common/header.go

```diff
@@ -13,8 +14,9 @@ var messageXid uint32 = 1
 
 func NewHeaderGenerator(ver int) func() Header {
 	return func() Header {
-		messageXid += 1
-		p := Header{uint8(ver), 0, 8, messageXid}
+		atomic.AddUint32(&messageXid, 1)
+		xid := atomic.LoadUint32(&messageXid)
```

Comment:
I believe the correct way to do this atomically is:
`xid := atomic.AddUint32(&messageXid, 1)`